Platform/ApollolakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c

```diff
@@ -555,6 +555,42 @@ RestoreOtgRole (
   MmioWrite32 (XhciBar + R_XHCI_MEM_DUAL_ROLE_CFG0, mOtgDualRoleCfg0);
 }
 
+/**
+  Set framebuffer range as WC using MTRR to improve performance.
+
+  The BSP MTRR needs to be programmed before FspSiliconInit() API so that
+  all APs' MTRRs will be syned up during FspSiliconInit() call.
+
+**/
+VOID
+SetFrameBufferWriteCombining (
+  VOID
+)
+{
+  UINT32             MsrIdx;
+  UINT32             MsrMax;
+  UINT32             Base;
+
+  // Enable Framebuffer as WC.
+  MsrMax = EFI_MSR_CACHE_VARIABLE_MTRR_BASE +
+           (2 * (UINT32)(AsmReadMsr64(EFI_MSR_IA32_MTRR_CAP) & B_EFI_MSR_IA32_MTRR_CAP_VARIABLE_SUPPORT));
+  for (MsrIdx = EFI_MSR_CACHE_VARIABLE_MTRR_BASE; MsrIdx < MsrMax; MsrIdx += 2) {
+    // Try to find a free MSR pair
+    if ((AsmReadMsr64(MsrIdx + 1) & B_EFI_MSR_CACHE_MTRR_VALID) == 0) {
+      break;
+    }
+  }
+
+  if (MsrIdx < MsrMax) {
+    // Framebuffer belongs to PMEM32 in PCI resource allocation.
+    // The 1st 256MB from PcdPciResourceMem32Base will be consumed by MEM32 resource.
+    // And framebuffer should be allocated to the next 256MB aligned address.
+    Base = (PcdGet32 (PcdPciResourceMem32Base) + SIZE_256MB)  &  ~(SIZE_256MB - 1);
+    AsmWriteMsr64 (MsrIdx,     Base | CACHE_WRITECOMBINING);
+    AsmWriteMsr64 (MsrIdx + 1, 0xF00000000ULL + B_EFI_MSR_CACHE_MTRR_VALID + (UINT32)(~(SIZE_256MB - 1)));
+  }
```

Comment:
Should we emit a warning message if a VMTRR could not be allocated?

Reply:
We can add a debug message for it.